clang/lib/CConv/CheckedRegions.cpp

```diff
@@ -187,6 +174,90 @@ bool CheckedRegionFinder::VisitCStyleCastExpr(CStyleCastExpr *E) {
   return true;
 }
 
+bool CheckedRegionFinder::VisitCallExpr(CallExpr *C) {
+  // TODO this should really be moved to the other visitor
+  // both for clarity and because it will allow it take advantage of
+  // redundancy checks
+  auto FD = C->getDirectCallee();
+  FoldingSetNodeID ID;
+  C->Profile(ID, *Context, true);
+  if (FD && FD->isVariadic()) {
+    Wild = true;
+  } else {
+    if (FD) {
+      auto type = FD->getReturnType();
+      Wild |= (!(FD->hasPrototype() || FD->doesThisDeclarationHaveABody()))
+        || containsUncheckedPtr(type)
+        || (any_of(FD->param_begin(), FD->param_end(), [this] (Decl *param) {
+              auto CVSet = Info.getVariable(param, Context);
+              return isWild(CVSet );
+            }));
+    }
+    handleChildren(C->children());
+    Map[ID] = Wild ? IS_UNCHECKED : IS_CHECKED;
+  }
+
+  return false;
+}
+
+bool CheckedRegionFinder::VisitVarDecl(VarDecl *VD) {
+  // Check if the variable is WILD.
+  auto CVSet = Info.getVariable(VD, Context);
+  // Check if the variable contains an unchecked type.
+  Wild = isWild(CVSet) || containsUncheckedPtr(VD->getType());
+  return true;
+}
+
+bool CheckedRegionFinder::VisitParmVarDecl(ParmVarDecl *PVD) {
+  // Check if the variable is WILD.
+  auto CVSet = Info.getVariable(PVD, Context);
+  Wild |= isWild(CVSet) | containsUncheckedPtr(PVD->getType());
+  return true;
+}
+
+bool CheckedRegionFinder::VisitMemberExpr(MemberExpr *E){
+  ValueDecl *VD = E->getMemberDecl();
+  if (VD) {
+    // Check if the variable is WILD.
+    std::set<ConstraintVariable *> CVSet = Info.getVariable(VD, Context);
+    for (auto Cv : CVSet) {
+      if (Cv->hasWild(Info.getConstraints().getVariables())) {
+        Wild = true;
+      }
+    }
+    // Check if the variable is a void*.
+    Wild |= containsUncheckedPtr(VD->getType());
+  }
+  return true;
+}
+
+bool CheckedRegionFinder::VisitDeclRefExpr(DeclRefExpr* DR) {
+  auto T = DR->getType();
+  auto D = DR->getDecl();
+  auto CVSet = Info.getVariable(D, Context);
+  bool IW = isWild(CVSet ) || containsUncheckedPtr(T);
+
+  if (auto FD = dyn_cast<FunctionDecl>(D)) {
```

Comment:
I'm wondering what situation this is handling. Is it when referring to a function pointer? If that's so, what will `containsUncheckedPtr(T)` do, above? Won't it see all the parameters and such, too?

Reply:
One is checking for function pointers (and regular calls) at the top level and using constraint info, the other handles buried function types. See the test `globalschecked.c`